### PR TITLE
feat(l1): in-memory cache database and new `EvmState`

### PR DESCRIPTION
**Pull request by xqft**
**Motivation**

implement a memory cache database and expand the `EvmState` struct into an enum. This is to take advantage of the block execution and validation implementations of the L1 to use it for proving in L2 mode.

**Description**
- adds new `EvmState` enum containing two variants (store and cache)
- adds `ExecutionDB` database for storing only the relevant data of the execution of a block (this is going to be passed into a zkVM for proving)
- adds and modifies some `EvmState` functions and adapts the codebase to those changes

### CODE REVIEW

File: crates/storage/cache/cache.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+use error::MemoryDBError;
+use revm::{
```
Comment (mpaulucci) [inline crates/storage/cache/cache.rs:2]: We don't want storage crates to have a dependency on `revm`, because we want to be able to switch evm implementations. So it should not leak outside the `evm` crate.

Comment (xqft) [reply crates/storage/cache/cache.rs:2]: got it, will move this to the `vm` crate, thanks!

File: crates/vm/vm.rs [hunk 2]
```diff
@@ -127,32 +136,66 @@ fn run_evm(
...
+                    .with_block_env(block_env)
+                    .with_tx_env(tx_env)
+                    .modify_cfg_env(|cfg| cfg.chain_id = chain_spec.chain_id)
+                    .with_spec_id(spec_id)
+                    .with_external_context(
+                        TracerEip3155::new(Box::new(std::io::stderr())).without_summary(),
+                    );
+                cfg_if::cfg_if! {
+                    if #[cfg(feature = "l2")] {
+                        use revm::{Handler, primitives::{CancunSpec, HandlerCfg}};
```
Comment (fmoletta) [inline crates/vm/vm.rs:152]: This looks like a lot of code repetition, maybe we could separate the setup of the evm builder in two steps so we could have something like
```
let mut evm_builder = match state {
         FirstVariant => Evm::builder().with_db()
         SecondVariant => Evm::builder().with_db()
};
let mut evm_builder = evm_builder.with_x().with_y();
```
We could also directly call build on the second statement

File: crates/vm/vm.rs [hunk 3]
```diff
@@ -235,113 +297,147 @@ fn run_without_commit(
...
+                    env.chain_id = chain_config.chain_id;
+                })
+                .build();
+            evm.transact().map_err(EvmError::from)?
+        }
+        EvmState::Execution(db) => {
+            let mut evm = Evm::builder()
+                .with_db(db)
+                .with_block_env(block_env)
+                .with_tx_env(tx_env)
```
Comment (fmoletta) [inline crates/vm/vm.rs:319]: Same suggestion here

File: crates/vm/vm.rs (hunk 2 above)
Comment (xqft) [reply crates/vm/vm.rs:152]: Yep I tried that but the problem is that `evm_builder`'s type has the DB as a generic (which can be a `Store` or `CacheDB<ExecutionDB>`) so each of the match's arms return a different type, which isn't allowed.

File: crates/vm/vm.rs (hunk 3 above)
Comment (xqft) [reply crates/vm/vm.rs:319]: Answered on the review above

File: crates/vm/vm.rs (hunk 2 above)
Comment (fmoletta) [reply crates/vm/vm.rs:163]: Oh, makes sense, then maybe we can add a generic helper function `build_evm` that takes a `T: Database` argument and returns an `Evm<T>`?

Comment (xqft) [reply crates/vm/vm.rs]: I missed a detail: the `Evm` and `EvmBuilder` have two and three generic types respectively, so I think such a function would be too complex for what it accomplishes.

I found another solution! the discrimination occurs at the `with_db()` call, so by moving it to the end of the chain we can minimize the duplicated code. I applied those changes in the latest commit.

File: crates/vm/execution_db.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            let account_info = store_wrapper
+                .basic(address)?
+                .ok_or(ExecutionDBError::NewMissingAccountInfo(address))?;
+            accounts.insert(address, account_info);
+
+            let account_storage = account_update
+                .added_storage
+                .iter()
+                .map(|(slot, value)| {
+                    let mut value_bytes = Vec::new();
```
Comment (fborello-lambda) [inline crates/vm/execution_db.rs:69]: We get a panic when using a `Vec`, a slice with just the length required works.
```suggestion
                    let mut value_bytes = [0u8; 32];
```

File: crates/vm/execution_db.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        let earliest_number = store_wrapper
+            .store
+            .get_earliest_block_number()?
+            .ok_or(ExecutionDBError::NewMissingBlockNumber())?;
+        let latest_number = store_wrapper
+            .store
+            .get_latest_block_number()?
+            .ok_or(ExecutionDBError::NewMissingBlockNumber())?;
+
+        let numbers = earliest_number..=latest_number;
```
Comment (mpaulucci) [inline crates/vm/execution_db.rs:90]: Why is this needed? This doesn't scale well, since for every new block you will have to do a new db read.

Comment (xqft) [reply crates/vm/execution_db.rs:90]: What you mean is that I'm reading the hashes of all blocks every time from the DB instead of retrieving them once and then adding each new hash to the list, right?

Comment (mpaulucci) [reply crates/vm/execution_db.rs:90]: In mainnet there are 21M blocks since genesis..

Comment (fborello-lambda) [reply crates/vm/execution_db.rs:90]: Yes, it's not optimized, but it’s not intended for the mainnet. We require all prior blocks **strictly necessary** to execute any desired block. The goal is to create the smallest possible "database", that can be de/serialiazable, to facilitate the `execute_block` operation for the prover. Since we're currently testing with L2 and have only a few blocks, this approach is feasible.

Comment (fborello-lambda) [reply crates/vm/execution_db.rs:90]: i've created an issue to track this problem/optimization.
#1016

Comment (mpaulucci) [reply crates/vm/execution_db.rs:90]: Ok

Comment (xqft) [reply crates/vm/execution_db.rs]: just applied that change.

File: crates/blockchain/payload.rs [hunk 6]
```diff
@@ -180,10 +180,14 @@ impl<'a> PayloadBuildContext<'a> {
         self.payload.header.number
     }
 
-    fn store(&self) -> &Store {
+    fn store(&self) -> Option<&Store> {
```
Comment (mpaulucci) [inline crates/blockchain/payload.rs:183]: would it make sense to return a Result instead of an Option so you can just use `?` from outside?

Comment (mpaulucci) [reply crates/blockchain/payload.rs:183]: is there really a case where this is None? It seems we end up adding a lot of defensive code for something that never happens

**Review by mpaulucci — APPROVED** [review]
